### PR TITLE
Globally unique rule identifiers

### PR DESCRIPTION
**Pull request by cipherboy**
#### Description:

Previously to this PR, rule directories were required to be unique within a guide. However, `rule_dir.json`'s format implicitly required rule identifiers to be unique across guides (i.e., globally unique rule identifiers). There was only a single instance of this in the entire product; a new identifier was assigned to the MacOS rule. Thus, I'm inclined to make the smaller of the two changes (renaming the one MacOS rule instead of restructuring all utilities which use `rule_dir.json`. 

This also causes `rule_dir_json.py` to fail when duplicate rule identifiers are encountered across products. 

Additionally, support to `build_product` for `macos1015` has been added. 

Thoughts?

### CODE REVIEW

**Review by yuumasato — COMMENTED** [review]
> Previously rule directories were required to be unique within a guide.

I'm not sure when this changed, do you have links?

**Comment by cipherboy** [comment]
@yuumasato said:
> @cipherboy said:
> > Previously rule directories were required to be unique within a guide.
>
> I'm not sure when this changed, do you have links?

Sorry, "previously to this PR..." :-)

In #3193 -- as a design point not really caught by its review and subsequent use elsewhere, `rule_dir.json` "unofficially" and "silently" lead to them needing to be globally unique.

Are you questioning whether rules (presently) need to be unique within a guide? E.g., if you can have `guide/one-group/rule_id` and `guide/two-group/rule_id`? I've tried before and generally the build system didn't like it at all -- I assumed it was generally known but perhaps not. Is there a way to fully specify the path of rules (e.g. `one-group/rule_id`) in the profile shorthand?

File: macos1015/profiles/moderate.profile [hunk 1]
```diff
@@ -28,5 +28,5 @@ description: |-
     consensus and release processes.
 
 selections:
-    - service_auditd_enabled
+    - service_com.apple.auditd_enabled
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply macos1015/profiles/moderate.profile:31]: There's already bad descriptions for MacOS though :D 

> ```yaml
> description: |-
>     The <tt>audit</tt> service is an essential userspace component of
>     the auditing system, as it is responsible for writing audit records to
>     disk.
>     {{{ describe_service_enable(service="auditd") }}}
> ```

`describe_service_enable` doesn't understand MacOS and `macos1015` doesn't define `init_system` (and it isn't `systemd` or `upstart` this gets expanded to the empty string.

```xml
          <xccdf-1.2:description xml:lang="en-US">The <html:code>audit</html:code> service is an essential userspace component of
the auditing system, as it is responsible for writing audit records to
disk.</xccdf-1.2:description>
```

Do we have an active maintainer for the MacOS content? There's only one profile... Is this a target for CaC in the future? Without a lot of work and contributions (there's two rules...) is there much point in keeping this content active? Looks like it was mostly Gabe's effort and from what I know, he's moved on from CaC... :)

Just a thought.

[the remark this answers is not in the document]
Comment (yuumasato) [reply macos1015/profiles/moderate.profile:31]: You have a point, content for MacOS has not been actively developed.
There is nothing explicit about this in the guidelines, but without active support and contributions, it feels natural to remove it.

**Comment by yuumasato** [comment]
> Are you questioning whether rules (presently) need to be unique within a guide? E.g., if you can have `guide/one-group/rule_id` and `guide/two-group/rule_id`? I've tried before and generally the build system didn't like it at all -- I assumed it was generally known but perhaps not. Is there a way to fully specify the path of rules (e.g. `one-group/rule_id`) in the profile shorthand?

Sorry I have not provided more context.
I asked because my understanding is that the rules in the project should be globally unique. 
I think having multiples rules with the same ID in different guides could lead to a lot of confusion.

**Comment by cipherboy** [comment]
@yuumasato 
> Sorry I have not provided more context.
> I asked because my understanding is that the rules in the project should be globally unique.
> I think having multiples rules with the same ID in different guides could lead to a lot of confusion.

Aha glad I asked :-) 

I don't know that I've ever seen any recommendations for global uniqueness. `docs/` doesn't mention anything matching `unique` `same.*id`... In `manual/developer/06_contributing_with_content.md` I see:

> A rule YAML file has one implied attribute:
>
> -   `id`: The primary identifier for the rule to be referenced from
>     profiles. This is inferred from the file name and links it to checks
>     and fixes with the same file name.

but nothing else really sticks out to me. Certainly until now, nothing has truly enforced it either.

I don't have a good justification for _requiring_ global uniqueness to be honest. :-) I took this path because it was the smallest change and affecting the least amount of people. The build system fully understands different guide directories --- so globally unique identifiers aren't necessary there (because product contains a guide reference...). I dunno. :-) 

I'd lean towards this change mostly because it doesn't seem that hard to enforce now and seems easier than changing `rule_dir.json`. If we wanted to make structural improvements to the builds system and use a JSON tree as a shortcut to walking the filesystem multiple times, I think we could improve `rule_dir.json` towards that goal, but otherwise, I don't know if it is worth it.

Thoughts?

**Comment by JAORMX** [comment]
/retest

AWS resource request issues

File: macos1015/profiles/moderate.profile (hunk 1 above)
[the remark this answers is not in the document]
Comment (cipherboy) [reply macos1015/profiles/moderate.profile]: @yuumasato Have y'all discussed keeping versus removing MacOS content? Should I drop these changes in favor of removing MacOS content right now?

**Comment by cipherboy** [comment]
@yuumasato @jan-cerny Could we get this considered for inclusion? Note the new top commit :-)

**Review by yuumasato — APPROVED** [review]
We discussed about this yesterday, and the general consensus is that having global unique rule IDs is the way to go.

For now we address the issue by renaming the macOS rule, neither the build system or the test suite are prepared to handle multiple rules with the same ID.

But in the future, if necessary, to address the the issue of disjoint "benchmark directories", the build system could be enhanced to support rule declaration in multiple files with disjoint `prodtypes` (This would make the rule easier to read if text for different products would be too different).

**Pull request merged**